### ISSUE TITLE
Best strategy to reduce compile times?

### ISSUE THREAD

**Issue by fverdugo** [gridap/Gridap.jl#723]
I have compared three main strategies to reduce compile time. I include here the results for the record.

Strategies:

0. Do nothing. Implemented in branch `master` commit https://github.com/gridap/Gridap.jl/commit/cd3b300836cac46ad8137588a2aa2ac881297e29

1. Introduce `@nospecialize` statements in assembly routines as suggested by @goerch. Implemented in branch https://github.com/gridap/Gridap.jl/tree/compilefast commit https://github.com/gridap/Gridap.jl/commit/8c051ddc86ce60b5171bdca46c7e654f65b3c900
1. Remove explicit `@inline` statements from several places of the code related with FE numerical integration. Implemented in branch https://github.com/gridap/Gridap.jl/tree/no_explicit_inline commit https://github.com/gridap/Gridap.jl/commit/b24fbaebba142f32ddaf58265206197239990f5c
1. Combination of 1. and 2. Implemented in branch https://github.com/gridap/Gridap.jl/tree/compilefast_and_no_explicit_inline commit https://github.com/gridap/Gridap.jl/commit/29259672bbb7b7b7c8447693f348b653075ed755

Test case: a MHD solver for the so-called Hunt benchmark (see below). 

Results (with Julia 1.6.1):

Strategy 0:
```
$ julia -O1 --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
199.817954 seconds (312.91 M allocations: 21.986 GiB, 4.56% gc time, 78.50% compilation time)
$ julia --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
526.765018 seconds (312.59 M allocations: 21.978 GiB, 1.80% gc time, 90.97% compilation time)
```

Strategy 1:
```
$ julia -O1 --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
185.388496 seconds (306.28 M allocations: 21.742 GiB, 4.89% gc time, 76.85% compilation time)
$ julia --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
319.738838 seconds (305.97 M allocations: 21.734 GiB, 2.88% gc time, 85.01% compilation time)
```

Strategy 2:
```
$ julia -O1 --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
174.756097 seconds (245.39 M allocations: 19.505 GiB, 4.38% gc time, 76.24% compilation time) 
$ julia --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
272.522085 seconds (245.03 M allocations: 19.492 GiB, 2.74% gc time, 85.57% compilation time)
```

Strategy 3:
```
$ julia -O1 --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
174.349611 seconds (245.14 M allocations: 19.487 GiB, 4.55% gc time, 76.23% compilation time)
$ julia --project=. -e 'include("hunt.jl"); @time TMP.hunt(5)'
272.425655 seconds (244.81 M allocations: 19.477 GiB, 2.91% gc time, 85.24% compilation time)
```

``` julia
# file hunt.jl
module TMP

using Gridap
using Gridap.Helpers
using Gridap.Algebra
using Gridap.CellData
using Gridap.ReferenceFEs
using Gridap.Geometry

function hunt(n)

  nx=n
  ny=n
  ν=1.0
  ρ=1.0
  σ=1.0
  B=VectorValue(0.0,1.0,0.0)
  f=VectorValue(0.0,0.0,1.0)
  L=1.0
  u0=1.0
  B0=norm(B)
  nsums = 10
  domain_phys = (-L,L,-L,L,0.0*L,0.1*L)

  # Reduced quantities
  Re = u0*L/ν
  Ha = B0*L*sqrt(σ/(ρ*ν))
  N = Ha^2/Re
  f̄ = (L/(ρ*u0^2))*f
  B̄ = (1/B0)*B
  α = 1.0
  β = 1.0/Re
  γ = N
  domain = domain_phys ./ L

  # Prepare problem in terms of reduced quantities
  map(x) = (2/sqrt(2))*VectorValue(
    sign(x[1])*(abs(x[1])*0.5)^0.5,
    sign(x[2])*(abs(x[2])*0.5)^0.5,
    x[3]*sqrt(2)/2)
  partition=(nx,ny,3)
  model = CartesianDiscreteModel(
    domain,partition;isperiodic=(false,false,true),map=map)
  Ω = Interior(model)
  labels = get_face_labeling(model)
  tags_u = append!(collect(1:20),[23,24,25,26])
  tags_j = append!(collect(1:20),[25,26])
  add_tag_from_tags!(labels,"noslip",tags_u)
  add_tag_from_tags!(labels,"insulating",tags_j)

  info = Dict{String,Float64}()
  # Reference FES
  k = 2
  T = Float64
  reffe_u = ReferenceFE(lagrangian,VectorValue{3,T},k)
  reffe_p = ReferenceFE(lagrangian,T,k-1;space=:P)
  reffe_j = ReferenceFE(raviart_thomas,T,k-1)
  reffe_φ = ReferenceFE(lagrangian,T,k-1)

  # Test spaces
  V_u = TestFESpace(Ω,reffe_u;dirichlet_tags="noslip")
  V_p = TestFESpace(Ω,reffe_p)
  V_j = TestFESpace(Ω,reffe_j;dirichlet_tags="insulating")
  V_φ = TestFESpace(Ω,reffe_φ;conformity=:L2)
  V = MultiFieldFESpace([V_u,V_p,V_j,V_φ])
  info["dofs u"] = num_free_dofs(V_u)
  info["dofs p"] = num_free_dofs(V_p)
  info["dofs j"] = num_free_dofs(V_j)
  info["dofs φ"] = num_free_dofs(V_φ)
  info["dofs total"] = num_free_dofs(V)

  # Trial Spaces TODO improve for parallel computations
  U_u = TrialFESpace(V_u,zero(VectorValue{3,T}))
  U_p = TrialFESpace(V_p)
  U_j = TrialFESpace(V_j,zero(VectorValue{3,T}))
  U_φ = TrialFESpace(V_φ)
  U = MultiFieldFESpace([U_u,U_p,U_j,U_φ])

  dΩ = Measure(Ω,2*k)

  conv(u,∇u) = (∇u')⋅u

  function a(x,dy)
    u, p, j, φ = x
    v_u, v_p, v_j, v_φ = dy
    ∫(
      β*(∇(u)⊙∇(v_u)) - p*(∇⋅v_u) -(γ*(j×B)⋅v_u) +
      (∇⋅u)*v_p +
      j⋅v_j - φ*(∇⋅v_j) - (u×B)⋅v_j +
      (∇⋅j)*v_φ ) * dΩ
  end

  function ℓ(dy)
    v_u, v_p, v_j, v_φ = dy
    ∫( v_u⋅f )*dΩ
  end

  function c(x,dy)
    u, p, j, φ = x
    v_u, v_p, v_j, v_φ = dy
    ∫( α*v_u⋅(conv∘(u,∇(u))) ) * dΩ
  end

  function dc(x,dx,dy)
    u, p, j, φ = x
    du , dp , dj , dφ  = dx
    v_u, v_p, v_j, v_φ = dy
    ∫( α*v_u⋅( (conv∘(u,∇(du))) + (conv∘(du,∇(u))) ) ) * dΩ
  end

  res(x,dy) = c(x,dy) + a(x,dy) - ℓ(dy)
  jac(x,dx,dy) = dc(x,dx,dy) + a(dx,dy)

  op=FEOperator(res,jac,U,V)
  solver = NLSolver(show_trace=true,method=:newton)
  xh = solve(solver,op)

  # Rescale quantities
  ūh,p̄h,j̄h,φ̄h = xh
  uh = u0*ūh
  ph = (ρ*u0^2)*p̄h
  jh = (σ*u0*B0)*j̄h
  φh = (u0*B0*L)*φ̄h
  grid_phys = UnstructuredGrid(get_grid(model))
  node_coords = get_node_coordinates(grid_phys)
  node_coords .= L .* node_coords
  Ω_phys = Gridap.Geometry.BodyFittedTriangulation(Ω.model,grid_phys,Ω.tface_to_mface)

  # Post process

  μ = ρ*ν
  grad_pz = -f[3]/ρ
  u(x) = analytical_hunt_u(L,L,μ,grad_pz,Ha,nsums,x)
  j(x) = analytical_hunt_j(L,L,σ,μ,grad_pz,Ha,nsums,x)

  title = "tmp"
  writevtk(Ω_phys,"$(title)_Ω_fluid",
    cellfields=[
    "uh"=>uh,"ph"=>ph,"jh"=>jh,"φh"=>φh,"u"=>u,"j"=>j,])

  k = 2
  dΩ_phys = Measure(Ω_phys,2*k)
  eu = u - uh
  ej = j - jh
  eu_h1 = sqrt(sum(∫( ∇(eu)⊙∇(eu) + eu⋅eu  )dΩ_phys))
  eu_l2 = sqrt(sum(∫( eu⋅eu )dΩ_phys))
  ej_l2 = sqrt(sum(∫( ej⋅ej )dΩ_phys))

  info["Re"] = Re
  info["Ha"] = Ha
  info["eu_h1"] = eu_h1
  info["eu_l2"] = eu_l2
  info["ej_l2"] = ej_l2

  info
end

function analytical_hunt_u(
  a::Float64,       # semi-length of side walls
  b::Float64,       # semi-length of Hartmann walls
  μ::Float64,       # fluid viscosity
  grad_pz::Float64, # presure gradient
  Ha::Float64,      # Hartmann number
  n::Int,           # number of sumands included in Fourier series
  x)                # evaluation point

  l = b/a
  ξ = x[1]/a
  η = x[2]/a

  V = 0.0; V0=0.0;
  for k in 0:n
    α_k = (k + 0.5)*π/l
    N = (Ha^2 + 4*α_k^2)^(0.5)
    r1_k = 0.5*( Ha + N)
    r2_k = 0.5*(-Ha + N)

    num = exp(-r1_k*(1-η))+exp(-r1_k*(1+η))
    den = 1+exp(-2*r1_k)
    V2 = (r2_k/N)*(num/den)

    num = exp(-r2_k*(1-η))+exp(-r2_k*(1+η))
    den = 1+exp(-2*r2_k)
    V3 = (r1_k/N)*(num/den)


    V += 2*(-1)^k*cos(α_k * ξ)/(l*α_k^3) * (1-V2-V3)
  end
  u_z = V/μ * (-grad_pz) * a^2

  VectorValue(0.0*u_z,0.0*u_z,u_z)
end


function analytical_hunt_j(
  a::Float64,       # semi-length of side walls
  b::Float64,       # semi-length of Hartmann walls
  σ::Float64,       # fluid conductivity
  μ::Float64,       # fluid viscosity
  grad_pz::Float64, # presure gradient
  Ha::Float64,      # Hartmann number
  n::Int,           # number of sumands included in Fourier series
  x)                # evaluation point

  l = b/a
  ξ = x[1]/a
  η = x[2]/a

  H_dx = 0.0; H_dy = 0.0
  for k in 0:n
    α_k = (k + 0.5)*π/l
    N = sqrt(Ha^2 + 4*α_k^2)
    r1_k = 0.5*( Ha + N)
    r2_k = 0.5*(-Ha + N)

    num = exp(-r1_k*(1-η)) - exp(-r1_k*(1+η))
    num_dy = exp(-r1_k*(1-η))*(r1_k/a) + exp(-r1_k*(1+η))*(r1_k/a)
    den = 1+exp(-2*r1_k)
    H2 = (r2_k/N)*(num/den)
    H2_dy = (r2_k/N)*(num_dy/den)

    num = exp(-r2_k*(1-η)) - exp(-r2_k*(1+η))
    num_dy = exp(-r2_k*(1-η))*(r2_k/a) + exp(-r2_k*(1+η))*(r2_k/a)
    den = 1+exp(-2*r2_k)
    H3 = (r1_k/N)*(num/den)
    H3_dy = (r1_k/N)*(num_dy/den)

    H_dx += -2*(-1)^k * sin(α_k * ξ)/(a*l*α_k^2) * (H2 - H3)
    H_dy += 2*(-1)^k * cos(α_k * ξ)/(l*α_k^3) * (H2_dy - H3_dy)
  end
  j_x = a^2*σ^0.5 / μ^0.5 * (-grad_pz) * H_dy
  j_y = a^2*σ^0.5 / μ^0.5 * (-grad_pz) * (-H_dx)

  VectorValue(j_x,j_y,0.0)
end

end # module

```

**Comment by fverdugo** [comment]
Conclusion: It seems that removing the `@inline` statements (Strategy 2) is the best choice.

I also prefer this than introducing `@nospecialize` since the latter leads to spurious allocations that grow linearly with the size of the assembly loop. 

@amartinhuertas can you check if the branch `no_explicit_inline` fixes your killer example?

If your killer example is fixed, I would merge strategy 2 into master.

**Comment by fverdugo** [comment]
cc @oriolcg

**Comment by amartinhuertas** [comment]
> Results (with Julia 1.6.1):

Why dont you update your local julia version to the latest version? :-)

**Comment by amartinhuertas** [comment]
might it make sense to downgrade `-O1` in GH CI actions?

**Comment by amartinhuertas** [comment]
> @amartinhuertas can you check if the branch no_explicit_inline fixes your killer example?

Checked. It fixes the killer example as well. (first call time latencies are comparable to those of @nospecialize)

**Comment by fverdugo** [comment]
Branch `no_explicit_inline` has been merged in PR #726 

Closing.
